### ISSUE TITLE
v0.13.5 of Flux breaks Flux.train! on a custom type with Zygote gradients

### ISSUE THREAD

**Issue by simonmandlik** [FluxML/Zygote.jl#1291]
Hi, in [Mill.jl](https://github.com/CTUAvastLab/Mill.jl) we have a special matrix type we call [PostImputingMatrix](https://github.com/CTUAvastLab/Mill.jl/blob/master/src/special_arrays/postimputing_matrix.jl), which fills in a vector trainable parameters ψ when they are missing in the input data. In short:

```julia
julia> using Flux, Mill

julia> A = Mill.PostImputingMatrix(reshape(1:9, 3, 3))
3×3 PostImputingMatrix{Int64, Base.ReshapedArray{Int64, 2, UnitRange{Int64}, Tuple{}}, Vector{Int64}}:
W:
 1  4  7
 2  5  8
 3  6  9

ψ:
 0
 0
 0

julia> B = Mill.maybehotbatch([1, missing, 3], 1:3)
3×3 MaybeHotMatrix with eltype Union{Missing, Bool}:
  true  missing    ⋅
   ⋅    missing    ⋅
   ⋅    missing   true

julia> A * B
3×3 Matrix{Int64}:
 1  0  7
 2  0  8
 3  0  9
```

For efficiency, this requires several some [custom `rrule`s](https://github.com/CTUAvastLab/Mill.jl/blob/23b5f66d6677d2cc3e15fec5f06ad8604732013c/src/special_arrays/postimputing_matrix.jl#L100-L119) as well as [`opt`ing out of default rules for Base.*](https://github.com/CTUAvastLab/Mill.jl/blob/23b5f66d6677d2cc3e15fec5f06ad8604732013c/src/special_arrays/postimputing_matrix.jl#L71).

When taking parameters and computing gradients, in `Flux@v0.13.4` it works as follows:

```julia
julia> Flux.params(A)
Params([PostImputingMatrix(W = [1 4 7; 2 5 8; 3 6 9], ψ = [0, 0, 0])])

julia> gradient(() -> sum(A*B), Flux.params(A)).grads
IdDict{Any, Any} with 2 entries:
  3×3 PostImputingMatrix => (W = [1 0 1; 1 0 1; 1 0 1], ψ = [1, 1, 1])
  :(Main.A)              => (W = [1 0 1; 1 0 1; 1 0 1], ψ = [1, 1, 1])
```

that is, `Flux.Params` contain the whole struct, `Zygote` returns structured gradients as well, and the only thing remaining for gradient descent is providing custom [`Flux.update!`](https://github.com/CTUAvastLab/Mill.jl/blob/23b5f66d6677d2cc3e15fec5f06ad8604732013c/src/special_arrays/special_arrays.jl#L100-L107).

In `Flux@v0.13.5` the same calls give:

```julia
julia> Flux.params(A)
Params([[1 4 7; 2 5 8; 3 6 9], [0, 0, 0]])

julia> gradient(() -> sum(A*B), Flux.params(A)).grads
IdDict{Any, Any} with 3 entries:
  :(Main.A)             => (W = [1 0 1; 1 0 1; 1 0 1], ψ = [1, 1, 1])
  [0, 0, 0]             => nothing
  [1 4 7; 2 5 8; 3 6 9] => nothing
```

`Flux.Params` no longer contain the whole struct and, consequently, gradients are not assigned to correct ids in the gradient Dict. This obviously will break many things such as updating parameters when training.

This is fixable by e.g. redefining `Flux.params` so that it behaves like in 0.13.4, or fixing the `gradient` call, which is probably a Zygote thing. But what is a preferable way to work with composite types like this one for the future?

I have posted an identical issue to `Zygote` as well, let me know and I will delete one.

**Comment by darsnack** [comment]
Closing as this is not the right repo for the issue and it is a dupe of https://github.com/FluxML/Flux.jl/issues/2045.
